### ISSUE TITLE
Back-port #762 to 5.x branch

### ISSUE THREAD

**Issue by lukebakken** [rabbitmq/rabbitmq-dotnet-client#767]
https://github.com/rabbitmq/rabbitmq-dotnet-client/pull/762

**Comment by lukebakken** [comment]
@danielmarbach PS `TaskCreationOptions.RunContinuationsAsynchronously` doesn't exist in `net451`, so #762 can't be back-ported.

**Comment by danielmarbach** [comment]
The create the same behavior of `TaskCreationOptions.RunContinuationsAsynchronously` would be 

```
Task.Run(() => tcs.TrySetResult());
```

but of course that has also implications when being done on the hotpath all the time. So from that perspective the semaphore version is probably better. Just wondering though if people running on 5.x version before Stebets patch have actually been running in any sort of problems being caused by using the TaskCompletionSource approach without `TaskCreationOptions.RunContinuationsAsynchronously` because all I've seen in the conversations was as statement on the PR like

> the prevent a common form of deadlocks

but no indication that it actually happened

**Comment by danielmarbach** [comment]
It is also possible to shim the behavior other than using `Task.Run` by doing this trickery

```
internal static class TaskCompletionSourceFactory {
#if NETFRAMEWORK
        static readonly FieldInfo StateField =
 typeof(Task).GetField("m_stateFlags", BindingFlags.NonPublic | BindingFlags.Instance);

#endif

		public static TaskCompletionSource<T> Create<T>(TaskCreationOptions options = TaskCreationOptions.None) {
#if NETFRAMEWORK
            //This lovely hack forces the task scheduler to run continuations asynchronously,
            //see https://stackoverflow.com/questions/22579206/how-can-i-prevent-synchronous-continuations-on-a-task/22588431#22588431
            var tcs = new TaskCompletionSource<T>(options);
            const int TASK_STATE_THREAD_WAS_ABORTED = 134217728;
            StateField.SetValue(tcs.Task, (int) StateField.GetValue(tcs.Task) | TASK_STATE_THREAD_WAS_ABORTED);
            return tcs;
#else
			return new TaskCompletionSource<T>(options | TaskCreationOptions.RunContinuationsAsynchronously);
#endif
		}
```

or the expression tree based version or if you want to go Marc Gravel kind of crazy you can do the IL version

https://github.com/StackExchange/StackExchange.Redis/blob/v1.2.x/StackExchange.Redis/StackExchange/Redis/TaskSource.cs

**Comment by lukebakken** [comment]
Thanks for your suggestions @danielmarbach . I think we'll leave `5.x` as-is and if issues are found it will be a good way to nudge people to version 6.

**Comment by lukebakken** [comment]
Fixed by #777
